Jenkins RT Pipeline

FernandoAndrade-NOAA opened the pull request:
This PR addresses #779. 
Three new files have been added into the /ci directory
There should be no baseline changes as this PR is for enhancement of testing functionality.
Allows for Jenkins to run regression tests on open PRs with the "Jenkins-run-RT" label on Hera and Orion

1. Jenkinsfile - instructions for the Jenkins Pipeline to run regression tests on supported machines (currently Hera and Orion)
2. rt-status.py - a utility script to summarize apparent changes in results for test cases after regression tests conclude, outputs to the Jenkins test run console
3. rt.sh - a refactored copy of the regression test script for the pipeline and developers to utilize with added options and use of the EPIC UPP test suite location for supported machines.
3a. from a developer's standpoint, starting regression tests may look like the following (ensuring to be synced with latest develop):
```
cd ci/
./rt.sh -a epic -r `pwd`/rundir -t `pwd`/../
```
The following options have been added to rt.sh:

- a - account to submit jobs with
- w - work directory where job scripts and their output logs will be contained, defaults to the current directory/work-upp
- h - home directory, where the test suite resides. Consists of the data input, output, job cards, timestamps, etc. and defaults to utilizing the EPIC UPP test suite directory for regression testing
- r - run directory, where the tests will be run along with their generated files to compare to current baselines in the home directory. defaults to various locations on machines specified in rt.sh, it may be preferable to adjust the location through this rt.sh option as shown above
- t - test_v, the local path to the UPP test version that will be tested before merging. Should typically be set to the example above for developers to run RTs
- b - git branch for cloning before testing
- u - git url for the PR branch that will be cloned
- c - enable git cloning 

Note: use of options b, u, and c will automatically set test_v to the cloned UPP location and will not typically be needed for developer testing as rt.sh will already be contained within the repository, they're legacy options for regression testing.

FernandoAndrade-NOAA committed:
change to epic hpc-stack

change location of hpc-stack to epic's and changed hpc to 1.2.0

FernandoAndrade-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

FernandoAndrade-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

FernandoAndrade-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

FernandoAndrade-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

FernandoAndrade-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

FernandoAndrade-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

FernandoAndrade-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

FernandoAndrade-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

FernandoAndrade-NOAA committed:
Merge branch 'NOAA-EMC:develop' into develop

FernandoAndrade-NOAA committed:
add jenkinsfile and rt-status scripts

FernandoAndrade-NOAA committed:
Update Jenkinsfile paths

FernandoAndrade-NOAA committed:
Update rt-status.py fix changed results and paths

FernandoAndrade-NOAA committed:
revert node_path cd

Commit:
add rt.sh and update paths

FernandoAndrade-NOAA committed:
re-enable tests

FernandoAndrade-NOAA committed:
update rundir paths

update run directory local paths to environment variable

FernandoAndrade-NOAA committed:
log and path cleanup

cleanup rt log for pushing through jenkins and cleanup paths in rt-status.py for rundir

FernandoAndrade-NOAA commented:
please ignore the review request from being taken out of draft

FernandoAndrade-NOAA committed:
update paths and .py inside rt.sh

update rundir paths and run rt-status.py from within rt.sh

FernandoAndrade-NOAA committed:
remove rundir print

FernandoAndrade-NOAA committed:
rt.sh save temp log file

FernandoAndrade-NOAA committed:
update log file names

FernandoAndrade-NOAA committed:
log refactor

move rt log cleanup into rt.sh and remove 3drtma debug print

FernandoAndrade-NOAA committed:
Merge branch 'NOAA-EMC:develop' into jenkins-ci

FernandoAndrade-NOAA committed:
Update rt.sh move rt log to /tests

FernandoAndrade-NOAA committed:
add hera 780 rt log

FernandoAndrade-NOAA commented:
@WenMeng-NOAA just an FYI, Hera logs have been pushed up with no changes as expected. Orion may need to be put on hold as the disk quota has been exceeded at the moment, so this may need to be pushed back until that's resolved.

WenMeng-NOAA commented:
@FernandoAndrade-NOAA The hera log looks good to me. Thanks for working on this.

FernandoAndrade-NOAA committed:
Merge branch 'NOAA-EMC:develop' into jenkins-ci

FernandoAndrade-NOAA committed:
orion RT log

FernandoAndrade-NOAA committed:
Update rt-status.py

FernandoAndrade-NOAA commented:
@WenMeng-NOAA I've run under an alternative workspace in /work2 to run the Orion tests, which should be fine in the meantime while work1's disk quota is resolved. Orion log is pushed and this should be ready for review/final testing. Please let me know if there's any desired changes and suggestions e.g. different default path values / additional log output (like the rt-status.py output). There should be no changed results. 

One thing to note is that the Jenkins RT Pipeline's configuration is currently hard coded to /work1, so there may still be disk quota hiccups after this is merged until that's fixed.

Thanks!

WenMeng-NOAA commented on ci/rt.sh:
```diff
@@ -0,0 +1,254 @@
+mac=$(hostname | cut -c1-1)
+mac2=$(hostname | cut -c1-2)
+mac3=$(hostname | cut -c1-4)
+if [ $mac = v -o $mac = m  ] ; then   # WCOSS_D (venus and mars)
```
@FernandoAndrade-NOAA Can you remove this if block for WCOSS_D since this platform was retired.

FernandoAndrade-NOAA replied:
Sure thing! With no explicit WCOSS2 check after removal of 64-67, I'm assuming you have an alternative test script on that machine you've been running RTs with?

WenMeng-NOAA replied:
@FernandoAndrade-NOAA Good catching. You might add settings for WCOSS2 as:
```
elif [ $mac = d -o $mac = c ]; then    # WCOSS2
 export machine=WCOSS2
 export homedir=/u/wen.meng/noscrub/ncep_post/post_regression_test_new
 export rundir=/lfs/h2/emc/ptmp/$USER
module load prod_util
```

FernandoAndrade-NOAA replied:
Added, thanks!

WenMeng-NOAA commented on ci/rt.sh:
```diff
@@ -0,0 +1,254 @@
+mkdir -p $workdir
+
+#differentiates for orion and hercules
+export rundir="${rundir}-upp-${machine}"
```
@FernandoAndrade-NOAA Do you think using "${rundir}/upp-machine" would make it easier for users to clean the temporary space periodically on R&D platforms if they run the UPP RTs by themselves? For example,  /scratch1/NCEPDEV/stmp2/Wen.Meng-upp-HERA  VS  /scratch1/NCEPDEV/stmp2/Wen.Meng/upp-HERA

FernandoAndrade-NOAA replied:
That may be a little better organized yes. I can go ahead and update that. The main purpose of this was just to avoid conflicts in the the shared filesystems. Would you have a preference for a different directory name or should /upp-${machine} be fine?

WenMeng-NOAA replied:
@FernandoAndrade-NOAA Your update looks good for me.

FernandoAndrade-NOAA committed:
Merge branch 'NOAA-EMC:develop' into jenkins-ci

FernandoAndrade-NOAA committed:
remove WCOSS_D check and update rundir

Removes the check and paths for retired WCOSS_D machine and updates rundir to include upp-${machine} as a subdirectory for organization.

FernandoAndrade-NOAA committed:
add WCOSS2 to rt.sh

WenMeng-NOAA commented on ci/rt.sh:
```diff
@@ -0,0 +1,255 @@
+ module load stack-intel-oneapi-mpi/2021.9.0
+ module load jasper/2.0.32
+ module load prod-util/1.2.2
+elif [ $mac = d -o $mac = c ]; then # WCOSS2
```
@FernandoAndrade-NOAA Forgot to mention that WCOSS2 is not using slurm job scheduler. So rt.sh failed on WCOSS2. You may remove this block. We working on that later.

FernandoAndrade-NOAA replied:
Sounds good, I'll remove the WCOSS2 lines in the meantime then, thanks for testing that.

FernandoAndrade-NOAA committed:
remove WCOSS2

New WCOSS2 lines will be temporarily removed due to the machine not utilizing slurm scheduler causing failures in the test script. It will be reintroduced at a later time.

WenMeng-NOAA commented:
The UPP RTs were successfully completed on WCOSS2. No baseline updates are needed in this PR.

WenMeng-NOAA commented:
@FernandoAndrade-NOAA My RTs on WCOSS2 was completed. Do you need to rerun tests on Hera and Orion? Otherwise, this PR looks good for me now.

FernandoAndrade-NOAA commented:
> @FernandoAndrade-NOAA My RTs on WCOSS2 was completed. Do you need to rerun tests on Hera and Orion? Otherwise, this PR looks good for me now.

I'll go ahead and do a final rerun just on Hera to ensure the path update for rundir doesn't break anything, but there should be no baseline changes.

FernandoAndrade-NOAA commented:
No issues in the Hera rerun, and no changes. This should be ready for merging. Thanks for your help @WenMeng-NOAA!

WenMeng-NOAA commented:
This PR is ready for merging.
